File: C02-Getting-Started/code/addbits.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Adds the two binary values together
 *
 * @param A char array A
 * @param B char array B
 * @param n length of A (same as B)
 * @return pointer to char array C (has length n + 1)
 */
char *addBits(char *A, char *B, int n)
{
    char r = 0;
    char *C = (char *) malloc(sizeof(char) * (n + 1));
    for(; (n - 1) >= 0; n--) {
        char s = A[n - 1] + B[n - 1] + r;
        C[n] = s % 2;
        r = s > 1 ? 1 : 0;
    }
    C[0] = r;
}

/**
 * Starting point of application
 */
int main()
{
    char A[] = {1, 0, 1, 1};
    char B[] = {1, 1, 0, 1};
    char *C = addBits(A, B, sizeof(A) / sizeof(char)); //should be 11000
    for(int i = 0; i < 5; i++)
        printf("%d", C[i]);

    return 0;
}

File: C02-Getting-Started/code/add_bits.py
def add_bits(bits_a, bits_b):
    r = 0
    bits_c = []
    for i in range(len(bits_a) - 1, -1, -1):
        s = bits_a[i] + bits_b[i] + r
        bits_c = [s % 2] + bits_c
        if s > 1:
            r = 1
        else:
            r = 0
    bits_c = [r] + bits_c
    return bits_c


b1 = [1, 0, 1, 1]
b2 = [1, 1, 0, 1]
print(add_bits(b1, b2))  # should be 11000
